lib/rules/no-unsupported-role-attributes.js

```diff
@@ -0,0 +1,84 @@
+import { roles, elementRoles } from 'aria-query';
+
+import AstNodeInfo from '../helpers/ast-node-info.js';
+import Rule from './_base.js';
+
+function createUnsupportedAttributeErrorMessage(attr, role, element) {
+  if (element) {
+    return `The attribute ${attr} is not supported by the element ${element} with the implicit role of ${role}`;
+  } else {
+    return `The attribute ${attr} is not supported by the role ${role}`;
+  }
+}
+function getImplicitRole(element, typeAttribute) {
+  if (element === 'input') {
+    for (let key of elementRoles.keys()) {
+      if (key.name === element) {
+        let attributes = key.attributes;
+        for (let attribute of attributes) {
+          if (attribute.name === 'type' && attribute.value === typeAttribute) {
+            return elementRoles.get(key)[0];
+          }
+        }
+      }
+    }
+  } else {
+    return elementRoles.get(elementRoles.keys().find((key) => key.name === element))[0];
+  }
+}
+
+export default class NoUnsupportedRoleAttributes extends Rule {
+  visitor() {
+    return {
+      ElementNode(node) {
```

Comment:
Maybe add a few more summary comments to explain what's going on at a high-level like: 

"First look to see if a role is explicitly defined"
"Otherwise check for an implicit role from the type"